### PR TITLE
Use `CONFIG_(BIG|LITTLE)_ENDIAN` instead of `__BYTE_ORDER__` for endianness checks

### PR DESCRIPTION
**Pull request by carlescufi**
See commit messages.

**Please Ignore the compliance check failure, it's just checkpatch doing its job with the new check**

Note that this PR includes the following fix: https://github.com/littlefs-project/littlefs/pull/742

### CODE REVIEW

File: drivers/w1/w1_net.c
Comment (gmarull) [inline drivers/w1/w1_net.c:186]: shouldn't this be using `IS_ENABLED`?

Comment (carlescufi) [reply drivers/w1/w1_net.c:186]: ooops, thanks!

File: modules/littlefs/zephyr_lfs_config.h
Comment (gmarull) [inline modules/littlefs/zephyr_lfs_config.h:152]: `&& defined(CONFIG_BIG_ENDIAN)`?

File: include/zephyr/toolchain.h
Comment (gmarull) [inline include/zephyr/toolchain.h:127]: just a question here, is this check needed anymore?

File: modules/littlefs/zephyr_lfs_config.h
Comment (carlescufi) [reply modules/littlefs/zephyr_lfs_config.h:152]: Hmm no, because it´s an `#elif`?

File: include/zephyr/toolchain.h
Comment (carlescufi) [reply include/zephyr/toolchain.h:127]: It's a double-check but I will defer to @tejlmand.

Comment (stephanosio) [reply include/zephyr/toolchain.h:127]: yes, as long as we use `__BYTE_ORDER__` and its friends in tree (i.e. right below).

Comment (carlescufi) [reply include/zephyr/toolchain.h:127]: Ah right, that's a good reason too :) but we could stop using them in the in-tree toolchain headers entirely if we wanted to (i.e. simplify even further), but I thought that was going too far.

File: drivers/w1/w1_net.c
Comment (stephanosio) [reply drivers/w1/w1_net.c]: I agree that attempting to support `__ORDER_PDP_ENDIAN__` is absurd. I have not seen a single architecture that uses `__ORDER_PDP_ENDIAN__` other than PDP-11, and I am pretty sure Zephyr is never going to run on PDP-11.

Comment (de-nordic) [reply drivers/w1/w1_net.c]: > I agree that attempting to support `__ORDER_PDP_ENDIAN__` is absurd. I have not seen a single architecture that uses `__ORDER_PDP_ENDIAN__` other than PDP-11, and I am pretty sure Zephyr is never going to run on PDP-11.

You are breaking my heart. Still got some hopes for VAX-11...

**Review by de-nordic — DISMISSED** [review]
mgmt/mcumgr is OK with the changes.

**Comment by de-nordic** [comment]
Btw. failing CI needs this fix https://github.com/zephyrproject-rtos/zephyr/pull/51656

**Comment by henrikbrixandersen** [comment]
Noting that we will likely still have lots of external code (modules) depending on the compiler definitions, what is the real gain here? Is this preparation for something else?

**Comment by henrikbrixandersen** [comment]
Also, if we intend to only allow one truth for endianess in-tree we ought to add a checkpatch check or usage of the compiler definitions will slowly creep back in.

**Comment by carlescufi** [comment]
> Noting that we will likely still have lots of external code (modules) depending on the compiler definitions, what is the real gain here? Is this preparation for something else?

The gain is to use Kconfig internally in-tree, because we have the Kconfig option. The fact that modules will continue to use `__BYTE_ORDER__` is not a reason not to be consistent ourselves, like we do in many other similar cases, eg. `__packed`.

**Comment by carlescufi** [comment]
> Also, if we intend to only allow one truth for endianess in-tree we ought to add a checkpatch check or usage of the compiler definitions will slowly creep back in.

Yep, I thought about that and I agree. @stephanosio any suggestions here?

**Comment by stephanosio** [comment]
> Yep, I thought about that and I agree. @stephanosio any suggestions here?

We can add a checkpatch check for it, something like:

https://github.com/zephyrproject-rtos/zephyr/blob/21c7e3ba65c20fa6cd574751780a010e8adb3dfa/scripts/checkpatch.pl#L6559-L6563

**Comment by carlescufi** [comment]
> > Yep, I thought about that and I agree. @stephanosio any suggestions here?
> 
> We can add a checkpatch check for it, something like:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/21c7e3ba65c20fa6cd574751780a010e8adb3dfa/scripts/checkpatch.pl#L6559-L6563

Thanks! Will add this to the PR

**Comment by cfriedt** [comment]
Just for clarification, we will still have `__BYTE_ORDER__`, `__ORDER_LITTLE_ENDIAN__`, and `__ORDER_BIG_ENDIAN__` defined by `toolchain.h`, correct?

It's just that I've spent the better part of a year, on and off, working on integrating Apache Thrift, and finally got this change upstream. It would be annoying if it were suddenly invalid.

https://github.com/apache/thrift/pull/2681/files

**Review by cfriedt — DISMISSED** [review]

**Comment by carlescufi** [comment]
> > > Yep, I thought about that and I agree. @stephanosio any suggestions here?
> > 
> > 
> > We can add a checkpatch check for it, something like:
> > https://github.com/zephyrproject-rtos/zephyr/blob/21c7e3ba65c20fa6cd574751780a010e8adb3dfa/scripts/checkpatch.pl#L6559-L6563
> 
> Thanks! Will add this to the PR

@henrikbrixandersen @stephanosio done, thanks!

**Review by Thalley — DISMISSED** [review]
ACK for BLE Audio

**Review by stephanosio — DISMISSED** [review]

**Review by rlubos — DISMISSED** [review]
+1 for networking

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
I would invert all of these `#ifndef CONFIG_BIG_ENDIAN ... #else ... #endif` sections to  `#ifdef CONFIG_BIG_ENDIAN ... #else ... #endif` in order to make the code more readable. I think the current approach with mechanically replacing makes the code much less logical/readable than before.

File: modules/littlefs/zephyr_lfs_config.h
Comment (gmarull) [reply modules/littlefs/zephyr_lfs_config.h:152]: Before:

```
if !LFS_NO_INTRINSICS && !CONFIG_BIG_ENDIAN
  return a;
elif !LFS_NO_INTRINSICS && CONFIG_BIG_ENDIAN
  return __builtin_bswap32(a);
else
  ...
```

Now:

```
if !CONFIG_BIG_ENDIAN
  return a;
elif !LFS_NO_INTRINSICS
  return __builtin_bswap32(a);
else:
 ...
```

If we now have: `LFS_NO_INTRINSICS && !CONFIG_BIG_ENDIAN`, we'll `return a`, whereas before the `else` branch would have been taken, and I think that is what is desired according to the description:

```c
// Builtin functions, these may be replaced by more efficient
// toolchain-specific implementations. LFS_NO_INTRINSICS falls back to a more
// expensive basic C implementation for debugging purposes
```

File: include/zephyr/toolchain.h
Comment (gmarull) [reply include/zephyr/toolchain.h:127]: If not defined, compiler will also complain about `__BYTE_ORDER__` not defined in the checks below, right? So, the message above doesn't tell me anything the compiler can't.

**Comment by carlescufi** [comment]
> I would invert all of these `#ifndef CONFIG_BIG_ENDIAN ... #else ... #endif` sections to `#ifdef CONFIG_BIG_ENDIAN ... #else ... #endif` in order to make the code more readable. I think the current approach with mechanically replacing makes the code much less logical/readable than before.

I disagree. Most platforms are little endian, and thus for me it makes sense to have the little endian version at the top, as the "default" option. I think it'd be confusing for people to have the big endian version of a bitfield section in a struct first, because nowadays people instinctively recognize the first block to be the common/default case, which is little endian.

**Comment by jhedberg** [comment]
> I disagree. Most platforms are little endian, and thus for me it makes sense to have the little endian version at the top, as the "default" option

I think it's just general good practice, that if you have an else-branch, that if possible, the if-condition should be testing for the existence of something rather than the non-existence of something. At least that computes much faster in my brain :) It doesn't really matter what the real world frequency of occurrence of the "something" behind the condition is. Just my two cents.. :)

**Comment by carlescufi** [comment]
> > I disagree. Most platforms are little endian, and thus for me it makes sense to have the little endian version at the top, as the "default" option
> 
> I think it's just general good practice, that if you have an else-branch, that if possible, the if-condition should be testing for the existence of something rather than the non-existence of something else. At least that computes much faster in my brain :) It doesn't really matter what the real world frequency of occurrence of the "something" behind the condition is. Just my two cents.. :)

I agree with that, but the question is whether we really want to have:

```
#ifdef CONFIG_BIG_ENDIAN
<big endian code that will almost never be used>
#else
<little endian code that will almost always be used>
#endif
```
I personally would not, but we are in the realm of personal preference here.
Perhaps it would help if, instead of `#ifndef`, I used `#if !defined()` ?

**Comment by henrikbrixandersen** [comment]
> I agree with that, but the question is whether we really want to have:
> 
> ```
> #ifdef CONFIG_BIG_ENDIAN
> <big endian code that will almost never be used>
> #else
> <little endian code that will almost always be used>
> #endif
> ```

That reads much more naturally - at least my somewhat twisted mind :) The less negations, the easier it is for me to grasp.

> I personally would not, but we are in the realm of personal preference here. Perhaps it would help if, instead of `#ifndef`, I used `#if !defined()` ? @jhedberg and @henrikbrixandersen

I don't think using `#if !defined()` improves the readability, no.

File: modules/canopennode/CO_driver_target.h
Comment (henrikbrixandersen) [inline modules/canopennode/CO_driver_target.h:46]: Take for instance this change. If `CONFIG_BIG_ENDIAN` is not defined, we define `CO_LITTLE_ENDIAN`. Otherwise we define `CO_BIG_ENDIAN`. This would read much naturally as:

```
#ifdef CONFIG_BIG_ENDIAN
#define CO_BIG_ENDIAN
#else
#define CO_LITTLE_ENDIAN
#endif
```

The same logic - in my mind - applies to all the other `#ifndef ... #else ... #endif` statements, but the above is a very literal example.

Comment (carlescufi) [reply modules/canopennode/CO_driver_target.h:46]: Maybe we could go case by case then. Since this is mostly about personal preference, I can change this particular one to match what you describe, and leave the other as `#if !defined()` unless a particular maintainer asks me to change it.

File: modules/littlefs/zephyr_lfs_config.h
Comment (gmarull) [reply modules/littlefs/zephyr_lfs_config.h:152]: It looks like the previous behavior was actually wrong, so please add a first commit fixing the issue.

**Comment by rlubos** [comment]
> I personally would not, but we are in the realm of personal preference here.

Totally. While it doesn't really make much difference to me, I'd slightly lean towards @henrikbrixandersen proposal, just feels more natural to see big endian stuff after seeing CONFIG_BIG_ENDIAN, it reads easier.

**Comment by cvinayak** [comment]
> I would invert all of these `#ifndef CONFIG_BIG_ENDIAN ... #else ... #endif` sections to `#ifdef CONFIG_BIG_ENDIAN ... #else ... #endif` in order to make the code more readable. I think the current approach with mechanically replacing makes the code much less logical/readable than before.

It is naturally (easily) readable when `if (true)` is followed by the true statements. And looking for big-endian code below `CONFIG_BIG_ENDIAN` is much easier without training the brain to understand the opposite as have to carefully read `#ifndef` or `#if !defined`.

My suggestion is to add a hidden
```
config LITTLE_ENDIAN
        # Hidden Kconfig option representing the default little-endian architecture
        bool
        depends on !BIG_ENDIAN
        default y
```

and use:
```
#if defined(CONFIG_LITTLE_ENDIAN)
/* Little-endian code */
#else
/* Big-endian code */
#endif
```

**Comment by carlescufi** [comment]
> My suggestion is to add a hidden
> 
> ```
> config LITTLE_ENDIAN
>         # Hidden Kconfig option representing the default little-endian architecture
>         bool
>         depends on !BIG_ENDIAN
>         default y
> ```

I thought about this, and so did @gmarull, but we thought that this was "cheating". @cvinayak is there a precedent for a hidden Kconfig like this that reflects the opposite of the visible one?

**Comment by henrikbrixandersen** [comment]
> I thought about this, and so did @gmarull, but we thought that this was "cheating". @cvinayak @Thalley @gmarull @henrikbrixandersen @rlubos is there a precedent for a hidden Kconfig like this that reflects the opposite of the visible one?

That sounds more like a Kconfig `choice` to me.

**Comment by stephanosio** [comment]
> My suggestion is to add a hidden
> 
> ```
> config LITTLE_ENDIAN
>         # Hidden Kconfig option representing the default little-endian architecture
>         bool
>         depends on !BIG_ENDIAN
>         default y
> ```
> 
> and use:

I agree that having both `CONFIG_LITTLE_ENDIAN` and `CONFIG_BIG_ENDIAN` is nicer and makes more sense.

Regarding its implementation, we should be using `choice`, however (`choice` will ensure mutual exclusivity between `CONFIG_LITTLE_ENDIAN` and `CONFIG_BIG_ENDIAN`).

**Comment by carlescufi** [comment]
> Regarding its implementation, we should be using `choice`, however (`choice` will ensure mutual exclusivity between `CONFIG_LITTLE_ENDIAN` and `CONFIG_BIG_ENDIAN`).

Will that break backwards compatibility though then? i.e. if users were setting `CONFIG_BIG_ENDIAN=y` out-of-tree?

**Comment by stephanosio** [comment]
> Will that break backwards compatibility though then? i.e. if users were setting `CONFIG_BIG_ENDIAN=y` out-of-tree?

No? The default will be `CONFIG_LITTLE_ENDIAN=y`. Setting `CONFIG_BIG_ENDIAN=y` will automatically set `CONFIG_LITTLE_ENDIAN=n` if implemented as a choice.

Something like:

```
choice
	default LITTLE_ENDIAN

config LITTLE_ENDIAN
	bool

config BIG_ENDIAN
	bool

endchoice
```

**Comment by carlescufi** [comment]
> > Will that break backwards compatibility though then? i.e. if users were setting `CONFIG_BIG_ENDIAN=y` out-of-tree?
> 
> No? The default will be `CONFIG_LITTLE_ENDIAN=y`. Setting `CONFIG_BIG_ENDIAN=y` will automatically set `CONFIG_LITTLE_ENDIAN=n` if implemented as a choice.
> 
> Something like:
> 
> ```
> choice
> 	default LITTLE_ENDIAN
> 
> config LITTLE_ENDIAN
> 	bool
> 
> config BIG_ENDIAN
> 	bool
> 
> endchoice
> ```

Right, sorry, my confusion. @henrikbrixandersen @rlubos you OK with this?

**Comment by henrikbrixandersen** [comment]
> Right, sorry, my confusion. @henrikbrixandersen @rlubos you OK with this?

Yes, but please use a named `choice` (perhaps `CONFIG_ENDIANNESS`?).

**Comment by tejlmand** [comment]
> Right, sorry, my confusion. @henrikbrixandersen @rlubos you OK with this?

@carlescufi  please do not change this into a choice.
The current `BIG_ENDIAN` is promptless and thereby cannot be selected by the user, but is selected by the arch or the board, for example:
https://github.com/zephyrproject-rtos/zephyr/blob/fd5fe6ca8e8fbc89263ad16f78403989e1308870/arch/Kconfig#L67

But a choice must have a prompt, and thus can suddenly be changed by the user.

Furthermore, other symbols cannot do a select on a choice symbol.

**Comment by stephanosio** [comment]
> But a choice must have a prompt, and thus can suddenly be changed by the user.

But that is not necessarily a bad thing? In fact, if you are working with a bi-endian architecture, users would want to be able to configure this.

> Furthermore, other symbols cannot do a select on a choice symbol.

This can easily be worked around by setting default for the choice in defconfig.

**Comment by nordicjm** [comment]
> This can easily be worked around by setting default for the choice in defconfig.

As part of boards v2, the defconfig fill will not be loaded at this point, so it cannot

**Comment by carlescufi** [comment]
@tejlmand thanks for the input. So then @cvinayak's [proposal](https://github.com/zephyrproject-rtos/zephyr/pull/51666#issuecomment-1293278549) to add a hidden options instead could work, right?

**Comment by nordicjm** [comment]
> Thanks @nordicjm! Do you have any objections against https://github.com/zephyrproject-rtos/zephyr/pull/51666#issuecomment-1293278549?

That looks good!

**Comment by carlescufi** [comment]
> > As part of boards v2, the defconfig fill will not be loaded at this point, so it cannot
> 
> well, then that sounds like something that needs to be fixed for v2.
> 
> p.s. arch symbols should not be selecting endianness.

I don´t want to speak for @nordicjm and @tejlmand, but board v2 is fundamentally different (and opt-in) in the way it determines architecture, so it might be that the `.defconfig` behavior here is intentional

**Comment by cfriedt** [comment]
> > Right, sorry, my confusion. @henrikbrixandersen @rlubos you OK with this?
> 
> Yes, but please use a named `choice` (perhaps `CONFIG_ENDIANNESS`?).

Yoda would say `CONFIG_ENDIAN_BIG` or `CONFIG_ENDIAN_LITTLE`... and then the choice could be `CONFIG_ENDIAN_CHOICE`.

Yoda

**Comment by tejlmand** [comment]
> But that is not necessarily a bad thing? In fact, if you are working with a bi-endian architecture, users would want to be able to configure this.

@stephanosio if you're working on such an architecture the architecture itself should define a choice which can then select the proper endianess based on user selection.

No reason to expose such a choice to user on other architectures where this is not relevant.

**Comment by tejlmand** [comment]
> > As part of boards v2, the defconfig fill will not be loaded at this point, so it cannot
> 
> well, then that sounds like something that needs to be fixed for v2.

I believe you're talking past each other.
The defconfig is part of Zephyr build Kconfig tree, not the SoC / arch trees.
So in a Zephyr build the defconfig is loaded, but in sysbuild (which can also use the arch / soc v2) there is no defconfig (as this file belongs to Zephyr).
Sysbuild has its own Kconfig tree, but shares part of it with Zephyr Kconfig tree.

But I guess further discussion of how this works is out of scope for this PR.

**Comment by tejlmand** [comment]
> @tejlmand thanks for the input. So then @cvinayak's [proposal](https://github.com/zephyrproject-rtos/zephyr/pull/51666#issuecomment-1293278549) to add a hidden options instead could work, right?

yes, and it would even allow arch to do `select LITTLE_ENDIAN` although they won't have to.

**Comment by henrikbrixandersen** [comment]
> > @tejlmand thanks for the input. So then @cvinayak's [proposal](https://github.com/zephyrproject-rtos/zephyr/pull/51666#issuecomment-1293278549) to add a hidden options instead could work, right?
> 
> 
> 
> yes, and it would even allow arch to do `select LITTLE_ENDIAN` although they won't have to.

Let's go with the two opposite options, then.

**Comment by stephanosio** [comment]
> @stephanosio if you're working on such an architecture the architecture itself should define a choice which can then select the proper endianess based on user selection.

So your suggestion is to add `CONFIG_ARM_BIG_ENDIAN`, `CONFIG_ARM_LITTLE_ENDIAN`, `CONFIG_SPARC_BIG_ENDIAN`, `CONFIG_SPARC_LITTLE_ENDIAN`, `CONFIG_MIPS_BIG_ENDIAN`, `CONFIG_MIPS_LITTLE_ENDIAN` and so on? Because that sounds far from ideal to me.

> No reason to expose such a choice to user on other architectures where this is not relevant.

I guess it is sort of a double edged sword -- using `choice` gives you "unified configuration" for endianness across all architectures that can be user-overridden; at the same time, it opens up the possibility for users to override this setting for the architectures/SoCs that do not support bi-endianness.

Anyways, this use case seems to be beyond the scope of this PR. Let us revisit this when the need arises (but yeah, the `SPARC` symbol should really not be selecting `BIG_ENDIAN`, that is just plain wrong and we should fix it).

> So in a Zephyr build the defconfig is loaded, but in sysbuild (which can also use the arch / soc v2) there is no defconfig (as this file belongs to Zephyr).

That sounds like a problem to me; but, I do admit I have not given enough thought on this matter as you have; so, I assume this can be worked around somehow.

**Comment by carlescufi** [comment]
@gmarull @cvinayak please take another look

File: arch/Kconfig
Comment (gmarull) [inline arch/Kconfig:175]: mixes tabs/spaces

File: modules/littlefs/zephyr_lfs_config.h
Comment (gmarull) [inline modules/littlefs/zephyr_lfs_config.h:178]: hmm this is also being changed, was it wrong?

Comment (carlescufi) [reply modules/littlefs/zephyr_lfs_config.h:178]: how strange, no. This is an oversight. Thanks!

**Review by gmarull — APPROVED** [review]

**Review by de-nordic — APPROVED** [review]
Mcumgr is OK.

**Review by Thalley — APPROVED** [review]
LE Audio ACK

File: include/zephyr/toolchain.h
Comment (stephanosio) [reply include/zephyr/toolchain.h:127]: > If not defined, compiler will also complain about `__BYTE_ORDER__` not defined in the checks below, right?

Unless you specify `-Wundef`, no.

Moreover, `-Wundef` is not enabled by default nor is it enabled by `-Wall`.

**Review by stephanosio — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

**Review by cvinayak — APPROVED** [review]
`subsys/bluetooth/controller/ll_sw` approved.

**Review by henrikbrixandersen — APPROVED** [review]
Much more readable, thanks!

File: scripts/checkpatch.pl
Comment (henrikbrixandersen) [inline scripts/checkpatch.pl:6568]:
```suggestion
			      "Use of the '$1' macro is disallowed. Use CONFIG_LITTLE_ENDIAN/CONFIG_BIG_ENDIAN instead\n" . $herecurr);
```

**Review by henrikbrixandersen — APPROVED** [review]
Minor improvement.

File: scripts/checkpatch.pl
Comment (carlescufi) [reply scripts/checkpatch.pl:6568]: Well spotted! will send a PR

**Comment by marc-hb** [comment]
Ironically, commit ae123c9bd1ea1 ("toolchain: Ensure that Kconfig endianness matches the compiler's") triggers an endianness sparse error. Reverting ae123c9bd1ea1 makes the error go away.

```shell
$ west build -p -b intel_adsp_cavs25 samples/hello_world/ -- -DSPARSE=y | grep error:

/home/mherber2/SOF/zephyr/include/zephyr/toolchain.h:133:2: error: "Endiannes mismatch"
/home/mherber2/SOF/zephyr/include/zephyr/toolchain.h:133:2: error: "Endiannes mismatch"
/home/mherber2/SOF/zephyr/include/zephyr/toolchain.h:133:2: error: "Endiannes mismatch"
/home/mherber2/SOF/zephyr/include/zephyr/toolchain.h:133:2: error: "Endiannes mismatch"
/home/mherber2/SOF/zephyr/include/zephyr/toolchain.h:133:2: error: "Endiannes mismatch"
/home/mherber2/SOF/zephyr/include/zephyr/toolchain.h:133:2: error: "Endiannes mismatch"
/home/mherber2/SOF/zephyr/include/zephyr/toolchain.h:133:2: error: "Endiannes mismatch"
/home/mherber2/SOF/zephyr/include/zephyr/toolchain.h:133:2: error: "Endiannes mismatch"
/home/mherber2/SOF/zephyr/include/zephyr/toolchain.h:133:2: error: "Endiannes mismatch"
/home/mherber2/SOF/zephyr/include/zephyr/toolchain.h:133:2: error: "Endiannes mismatch"
/home/mherber2/SOF/zephyr/include/zephyr/toolchain.h:133:2: error: "Endiannes mismatch"
/home/mherber2/SOF/zephyr/include/zephyr/toolchain.h:133:2: error: "Endiannes mismatch"
/home/mherber2/SOF/zephyr/include/zephyr/toolchain.h:133:2: error: "Endiannes mismatch"
/home/mherber2/SOF/zephyr/include/zephyr/toolchain.h:133:2: error: "Endiannes mismatch"
/home/mherber2/SOF/zephyr/include/zephyr/toolchain.h:133:2: error: "Endiannes mismatch"
/home/mherber2/SOF/zephyr/include/zephyr/toolchain.h:133:2: error: "Endiannes mismatch"
/home/mherber2/SOF/zephyr/include/zephyr/toolchain.h:133:2: error: "Endiannes mismatch"
/home/mherber2/SOF/zephyr/include/zephyr/toolchain.h:133:2: error: "Endiannes mismatch"

```

This could be a problem with sparse itself, I haven't investigated this _at all_, only  bisected it. Just a heads-up.

cc: @kv2019i , @juimonen , @lyakh , @lgirdwood

**Pull request merged**
